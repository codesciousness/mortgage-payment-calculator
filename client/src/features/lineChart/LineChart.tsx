import React from 'react';
import { Chart } from 'react-google-charts';
import './LineChart.css';
import { selectAmortizationSchedule, AmortizationDetail } from '../loansSlice';
import { useAppSelector } from '../../app/hooks';
import { stringToNum } from '../../util/calculations';

type ColumnHeaders = [string, string, string, string];

type DataPoints = [string, number, number, number];

const LineChart = (): JSX.Element => {
    const amortizationSchedule = useAppSelector(selectAmortizationSchedule);

    const columnHeaders: ColumnHeaders = ['Date', 'Principal', 'Interest', 'Balance'];
    const dataPoints: DataPoints = amortizationSchedule.map((row: AmortizationDetail) => 
    [row.date, stringToNum(row.totalPrincipal), stringToNum(row.totalInterest), stringToNum(row.remainingBalance)]);
    const data = [columnHeaders, ...dataPoints];

    const options = {
        curveType: 'function',
        legend: 'none',
        colors: [
            'PaleTurquoise',
            'DeepSkyBlue',
            'SteelBlue'
        ],
        chartArea: {
            width: '80%',
            height: '90%'
        },
        hAxis: {
            showTextEvery: 120,
            gridlines: {
                color: 'LightGray'
            }
        },
        vAxis: {
            format: 'short',
            gridlines: {
                multiple: 50000
            },
            minorGridlines: {
                count: 0
            }
        }
    };

    return (
        <Chart
            chartType='LineChart'
            width='100%'
            height='400px'
            data={data}
            options={options}
        />
    );
};

export default LineChart;