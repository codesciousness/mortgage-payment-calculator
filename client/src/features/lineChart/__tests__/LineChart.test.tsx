import React from 'react';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import { create } from 'react-test-renderer';
import { store } from '../../../app/store';
import LineChart from '../LineChart';

jest.mock('react-google-charts', () => ({Chart: () => <section data-testid="Chart">Chart</section>}));

describe('<LineChart />', () => {
    test('LineChart component is defined', () => {
        expect(<LineChart/>).toBeDefined();
    });
    test('component matches the snapshot', () => {
        const component = create(
            <Provider store={store}>
                <LineChart/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders a Chart component', () => {
        render(
            <Provider store={store}>
                <LineChart/>
            </Provider>
        );
        expect(screen.getByTestId('Chart')).toBeInTheDocument();
        expect(screen.getByText('Chart')).toBeInTheDocument();
    });
});