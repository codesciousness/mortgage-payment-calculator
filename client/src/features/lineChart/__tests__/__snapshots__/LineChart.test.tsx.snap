// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LineChart /> component matches the snapshot 1`] = `
<section
  className="LineChart"
  id="LineChart"
>
  <section
    data-testid="Chart"
  >
    Chart
  </section>
</section>
`;
